### PR TITLE
Support for functional localization

### PR DESCRIPTION
**Pull request by BKHMSI**
Users can now perform functional localization as described in [Brain-Like Language Processing via a Shallow Untrained Multihead Attention Network](https://arxiv.org/abs/2406.15109)

Changes:
- Localization stimuli saved in `data/fedorenko2010_localization` and can be loaded via the `data_registry`
- Localization script can be found in `model_helpers/localize` that computes language mask according to the paper mentioned above
- Language mask is cached in `.brainio`
- `HuggingfaceSubject` class was adapted to extract activations from multiple layers at once and make use of the localization script if the `use_localizer` flag is set to True. This extracts only the language selective units from all the activations.

Usage:
- Example script can be found in `examples/score_localization`

```py
benchmark = load_benchmark('Pereira2018.243sentences-linear')

num_blocks = 12
layer_names = [f'transformer.h.{block}.{layer_type}' 
    for block in range(num_blocks) 
    for layer_type in ['ln_1', 'attn', 'ln_2', 'mlp']
]

model = HuggingfaceSubject(model_id='gpt2', 
    region_layer_mapping={ArtificialSubject.RecordingTarget.language_system: layer_names},
    use_localizer=True,
    localizer_kwargs={
        'hidden_dim': 768,
        'batch_size': 16,
        "top_k": 4096,
    }
)

model_score = benchmark(model)
```

### CODE REVIEW

**Review by mschrimpf — APPROVED** [review]

File: brainscore_language/model_helpers/localize.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+BRAINIO_CACHE = os.environ.get("BRAINIO", f"{Path.home()}/.brainio")
+os.environ["TOKENIZERS_PARALLELISM"] = "False"
+
+logger = logging.getLogger(__name__)
+
+# Code adapted from: https://github.com/bkhmsi/brain-language-suma
+
+class Fed10_langlocDataset(Dataset):
+    def __init__(self):
+        self.num_samples = 240
```
[the remark this answers is not in the document]
Comment (BKHMSI) [reply brainscore_language/model_helpers/localize.py:28]: line #103 in the `extract_representations` function

File: brainscore_language/data/fedorenko2010_localization/__init__.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    paths = glob(f"{Path(__file__).parent }/*.csv")
+    data = pd.read_csv(paths[0])
+    for path in paths[1:]:
+        run_data = pd.read_csv(path)
+        data = pd.concat([data, run_data])
+
+    data["sent"] = data["stim2"].apply(str.lower)
+
+    for stimuli_idx in range(3, 14):
+        data["sent"] += " " + data[f"stim{stimuli_idx}"].apply(str.lower)
```
[the remark this answers is not in the document]
Comment (BKHMSI) [reply brainscore_language/data/fedorenko2010_localization/__init__.py]: added

File: brainscore_language/model_helpers/localize.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+
+from glob import glob
+from tqdm import tqdm
+from torch.utils.data import Dataset, DataLoader
+from pathlib import Path
+
+from brainscore_language import load_dataset
+
+BRAINIO_CACHE = os.environ.get("BRAINIO", f"{Path.home()}/.brainio")
+os.environ["TOKENIZERS_PARALLELISM"] = "False"
```
[the remark this answers is not in the document]
Comment (BKHMSI) [reply brainscore_language/model_helpers/localize.py]: added

File: brainscore_language/model_helpers/localize.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mschrimpf) [reply brainscore_language/model_helpers/localize.py:28]: ok I'm not actually sure what this does -- looks like it's just used to zero-fill `layer_name` (??)
Could this not also be derived from `self.sentences`?
```
   final_layer_representations = {
        "sentences": {layer_name: np.zeros((langloc_dataset.num_samples, hidden_dim)) for layer_name in layer_names},
        "non-words": {layer_name: np.zeros((langloc_dataset.num_samples, hidden_dim)) for layer_name in layer_names}
    }
```

[the remark this answers is not in the document]
Comment (BKHMSI) [reply brainscore_language/model_helpers/localize.py]: replaced `langloc_dataset.num_samples` with `len(langloc_dataset.sentences)`

File: .gitignore [hunk 4]
```diff
@@ -136,7 +136,7 @@ dmypy.json
 
 ### project specific additions:
 
-brainscore_language/data
+# brainscore_language/data
```
Comment (mschrimpf) [inline .gitignore:139]: ?

File: brainscore_language/metrics/rdm/__init__.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from brainscore_language import metric_registry
```
Comment (mschrimpf) [inline brainscore_language/metrics/rdm/__init__.py:1]: could we just import this metric from `brain-score-vision`?

**Review by mschrimpf — APPROVED** [review]
looks good; but please check comments

File: .gitignore (hunk 4 above)
Comment (BKHMSI) [reply .gitignore]: uncommented in the new commit

File: brainscore_language/metrics/rdm/__init__.py (hunk 5 above)
Comment (BKHMSI) [reply brainscore_language/metrics/rdm/__init__.py:1]: that would require `brain-score-vision` to be a dependency for `brain-score-language`, not sure if that's a good idea?

**Comment by mschrimpf** [comment]
@BKHMSI OK to merge?

**Comment by BKHMSI** [comment]
Yes, it was tested using Python 3.11
